SLEP018 Pandas output for transformers with set_output (#68)
scikit-learn/enhancement_proposals: slep018/proposal.rst

.. _slep_018:

=======================================================
SLEP018: Pandas Output for Transformers with set_output
=======================================================

:Author: Thomas J. Fan
:Status: Draft
:Type: Standards Track
:Created: 2022-06-22

Abstract
--------

> This SLEP proposes a ``set_output`` method to configure the output container of

Reviewer: ```suggestion
This SLEP proposes a ``set_output`` method to configure the output data container of
```

scikit-learn transformers.

Detailed description
--------------------

Currently, scikit-learn transformers return NumPy ndarrays or SciPy sparse
matrices. This SLEP proposes adding a ``set_output`` method to configure a
transformer to output pandas DataFrames::

   scalar = StandardScalar().set_output(transform="pandas")
   scalar.fit(X_df)

   # X_trans_df is a pandas DataFrame
   X_trans_df = scalar.transform(X_df)

> The index of the output DataFrame must match the index of the input. If the

Reviewer 2: How is sparse data treated? If I do ``set_output(transform="pandas")`` on a ``OneHotEncoder(sparse=True)``, will it error or will it produce a dense pandas array? How about estimators that don't have an explicit dense support, like CountVectorizer?
For ``CountVectorizer``, erroring seems natural, for OHE it seems strange to require the user to set the output intention in two places.

Author: > If I do set_output(transform="pandas") on a OneHotEncoder(sparse=True), will it error or will it produce a dense pandas array? How about estimators that don't have an explicit dense support, like CountVectorizer?

Error in both cases.

> for OHE it seems strange to require the user to set the output intention in two places.

I think it's strange not to error. If `sparse=True` and a dense pandas array is returned, then it is not consistent with `sparse=True`. As long as pandas + sparse is not supported, I prefer the explicitness of configuring in two places.

From a workflow point of view, the OHE will be in a pipeline, so it will end up to be one extra configuration (`sparse=True`). This is because `set_output` will end up being called on pipeline which configures everything:

```python
preprocessor = ColumnTransformer([("cat", OneHotEncoder(sparse=True), ...])
pipeline = make_pipeline(preprocessor, ...)
pipeline.set_output(transform="pandas")
```

Reviewer 2: You're right, in a pipeline it's not so bad. It might be worth calling out in the OHE documentation or in the error message? I agree it's better to be explicit, but it might also be surprising to users who don't know that the default is `sparse=True`, in particular because it might be invisible if used in a column transformer.

transformer does not support ``transform="pandas"``, then it must raise a
``ValueError`` stating that it does not support the feature.

> For this SLEP, ``set_output`` will only configure the output for dense data.  If
> the transformer returns sparse data, then ``transform`` will raise a
> ``ValueError`` if ``set_output(transform="pandas")``.

Reviewer: ```suggestion
This SLEP's only focus is dense data for ``set_output``. If a transformer returns
sparse data, e.g. `OneHotEncoder(sparse=True), then ``transform`` will raise a
``ValueError`` if ``set_output(transform="pandas")``. Dealing with sparse output
might be the scope of another future SLEP.
```

For a pipeline, calling ``set_output`` on the pipeline will configure all steps
in the pipeline::

   num_prep = make_pipeline(SimpleImputer(), StandardScalar(), PCA())
   num_preprocessor.set_output(transform="pandas")

   # X_trans_df is a pandas DataFrame

>    X_trans_df = num_preprocessor.fit_transform(X_df)
> 

Reviewer: ```suggestion
   X_trans_df = num_preprocessor.fit_transform(X_df)
   
   # X_trans_df is again a pandas DataFrame
   X_trans_df = num_preprocessor[0].transform(X_df)

```

Meta-estimators that support ``set_output`` are required to configure all inner
transformer by calling ``set_output``. If an inner transformer does not define
``set_output``, then an error is raised.

Global Configuration
....................

> This SLEP proposes a global configuration flag that sets the output for all

Reviewer: ```suggestion
For ease of use, this SLEP proposes a global configuration flag that sets the output for all
```

transformers::

   import sklearn
   sklearn.set_config(transform_output="pandas")

The global default configuration is ``"default"`` where the transformer
determines the output container.

Implementation
--------------

> The implementation of this SLEP is in :pr:`23734`.

Reviewer: ```suggestion
A possible implementation of this SLEP is worked out in :pr:`23734`.
```
I don't want to implicitly vote for the PR when voting on the SLEP, even if merging this PR is highly likely.

Backward compatibility
----------------------

There are no backward compatibility concerns, because the ``set_output`` method

> is a new API. Third party transformers can opt-in to the API by defining

Reviewer 2: Do they have to opt-in to respect the global flag?

Reviewer: There is no backward compatibility concern as long as we don't change the default of `sklearn.set_config(transform_output=XXX)`, correct?

Author: There isn't a backward compatibility concern, but there is an issue around if a third party estimator should respect the global flag. Concretely:

```python
sklearn.set_config(transform_output="pandas")

# Should we require this to be a dataframe?
third_party_transformer.transform(X_df)
```

I'm leading toward letting the library decide if it wants to respect the global configuration.

Reviewer 2: Exactly that was the case that seemed underspecified to me. I'm ok to leave it up to the library, which means we won't add it to the common tests that this errors out if it's not supported. Not sure if it's worth adding that to the doc as a sentence/half-sentence? Otherwise the doc doesn't really tell the third party estimator authors what they should be doing.

``set_output``.

Alternatives
------------

Alternatives to this SLEP includes:

1. `SLEP014 <https://github.com/scikit-learn/enhancement_proposals/pull/37>`__
   proposes that if the input is a DataFrame than the output is a DataFrame.
2. :ref:`SLEP012 <slep_012>` proposes a custom scikit-learn container for dense

>    and sparse data that contains feature names. This SLEP also proposes a custom

Reviewer 2: The sparse data is now in future work, right?

   container for sparse data, but pandas for dense data.
3. Prototype `#20100
   <https://github.com/scikit-learn/scikit-learn/pull/20100>`__ showcases
   ``array_out="pandas"`` in `transform`. This API is limited because does not
   directly support fitting on a pipeline where the steps requires data frames
   input.

Discussion
----------

A list of issues discussing Pandas output are: `#14315
<https://github.com/scikit-learn/scikit-learn/pull/14315>`__, `#20100
<https://github.com/scikit-learn/scikit-learn/pull/20100>`__, and `#23001
<https://github.com/scikit-learn/scikit-learn/issueas/23001>`__.

Future Extensions
-----------------

> 

Reviewer: ```suggestion
For information only!
```

Sparse Data
...........

> The Pandas DataFrame is not suitable to provide column names because it has

Reviewer 2: ```suggestion
The Pandas DataFrame is not suitable to provide column names for sparse data because it has
```

.. _slep_018:

=======================================================
SLEP018: Pandas Output for Transformers with set_output
=======================================================

:Author: Thomas J. Fan
:Status: Draft
:Type: Standards Track
:Created: 2022-05-23

Abstract
--------

This SLEP proposes a ``set_output`` method to configure the output container of
scikit-learn transformers.

Detailed description
--------------------

Currently, scikit-learn transformers return NumPy ndarrays or SciPy sparse matrices.
This SLEP proposes adding a ``set_output`` method to configure a transformer to output
pandas DataFrames::

   scalar = StandardScalar().set_output(transform="pandas")
   scalar.fit(X_df)

   # X_trans_df is a pandas DataFrame
   X_trans_df = scalar.transform(X_df)

The index of the output DataFrame must match the index of the input. For this
SLEP, ``set_output`` will only configure the output for dense data. If a
transformer returns sparse data, then ``transform`` will error if ``set_output``

> is to "pandas". If a transformer always returns sparse data, then calling

Reviewer 3: ```suggestion
is set to "pandas". If a transformer always returns sparse data, then calling
```

`set_output="pandas"` may raise an error.

> For a pipeline, calling ``set_output`` on the pipeline will configure all steps in the

Reviewer 3: Are there meaningful cases with an intermediate sparse representation (which would not really call for sparse data with column names)???

pipeline::

   num_prep = make_pipeline(SimpleImputer(), StandardScalar(), PCA())
   num_preprocessor.set_output(transform="pandas")

   # X_trans_df is a pandas DataFrame
   X_trans_df = num_preprocessor.fit_transform(X_df)

> Meta-estimators that support ``set_output`` are required to configure all estimators

Reviewer 3: All estimators? All transformers?

Author: In the context of this SLEP, it does make sense to reduce the scope to "all transformers". I was thinking about a future where we had `set_output(predict="pandas")` for non-transformers.

> by calling ``set_output``.

Reviewer 3: What is the procedure when this attribute is unavailable on a child estimator? Presumably the child should raise a `ValueError` if the value "pandas" is not supported. What should the parent do then?

Author: If `set_output` is not defined for the child estimator, then the parent errors. If the child does not support "pandas", then the child should error.

In both cases, the parent is not properly configured because one of the children failed the `set_output` call. For a user point of view, they can not use `set_output` for their meta-estimator.

Reviewer 3: Should this case be in the slep?

Author: It is in the SLEP here: https://github.com/thomasjpfan/enhancement_proposals/blob/218b76ad5612663ab9408dd4d824df357dda4d05/slep018/proposal.rst?plain=1#L76-L78 but it makes sense to move it up to this section.
